wayland-protocols/mir-shell-unstable-v1.xml

```diff
@@ -0,0 +1,533 @@
+    representations about the suitability of this software for any
+    purpose.  It is provided "as is" without express or implied
+    warranty.
+
+    THE COPYRIGHT HOLDERS DISCLAIM ALL WARRANTIES WITH REGARD TO THIS
+    SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND
+    FITNESS, IN NO EVENT SHALL THE COPYRIGHT HOLDERS BE LIABLE FOR ANY
+    SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
+    AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION,
+    ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF
+    THIS SOFTWARE.
+  </copyright>
+
+  <interface name="mir_shell_v1" version="1">
+    <description summary="create surface archetypes that are recognised by the desktop">
+      Clients can use this interface to assign an archetype to wl_surfaces.
+
+      An archetype is like a `wl_surface` role: a surface may have at most one
+      archetype, but unlike a `wl_surface` role a client may assign a new archetype
+      to a surface which already has one. This will remove the old archetype and
+      apply the new one atomically.
+
+      The archetype of a `wl_surface` influences the window-management policies
+      applied to it.
+
+      All archetype state is double-buffered; changing a surface archetype or
+      updates to any archetype state are not applied until the `wl_surface` is
+      `commit`ted.
+
+      Typically, archetype state will be used to augments `xdg_toplevel` state. If a
+      client intends to use an archetype, it SHOULD assign the surface an initial
+      archetype during the initial commit without a buffer attached (see `xdg_surface`).
+
+      Any events sent to an archetype object are latched to and extend the
+      `xdg_surface.configure` event. Any such events should be treated as a part of
+      an atomic set of configuration changes (including any `xdg_toplevel` events)
+      with the `xdg_surface.configure` event committing the accumulated state and
+      requiring a `xdg_surface.ack_configure` call, as normal.
+
+      Changing archetype follows a similar sequence to the initial `xdg_surface`
+      commit. First, a new archetype must be committed. This must be the only state
+      committed in the `wl_surface.commit` request. The compositor will respond with
+      any changes to the surface state prompted by the change of archetype followed
+      by an `xdg_surface.configure` event. The client must ack the configure event
+      (as normal) and the subsequent `wl_surface.commit` will cause the new
+      archetype to be fully applied.
+
+      There are some `wl_surface` roles that conflict with the archetypes described
+      here. Attempting to set both an archetypes and such roles on a surface is a
+      protocol error. In particular, `wl_subsurface`, `wl_cursor` or `xdg_popup` may
+      not be combined with an archetype.
```

Comment:
Since the behaviour of “you tried a bad combination” is “your application crashes” and there's no other way than reading the protocol of discovering what combinations are bad I don't think we can be anything but exhaustive in enumerating the valid combinations here.

What `wl_surface` roles *should* be supported? `xdg_toplevel`, and…? *Maybe* layer shell surfaces, but they're already pretty special and I don't think any of this window management applies to them. `wl_shell` is dead dead dead and no client is going to use it in new code.

This does further suggest that the correct object type for this extension to act on is `xdg_toplevel` rather than `wl_surface`.

Reply:
> This does further suggest that the correct object type for this extension to act on is `xdg_toplevel` rather than `wl_surface`.

I don't, at this stage, want to tie `mir-shell` to `xdg-shell`. As specified here it is compatible with `wl-shell`. I know `wl-shell` is deprecated, but I want to be open to a hypothetical alternative (for AR perhaps?) and it still exists.

In future we will likely want to propose something based on experience to wayland-protocols, that is the time to think more closely about integration with (or extension of) `xdg-shell`.

Relying on `xdg_toplevel` and `xdg_positioner` is clearly a plausible change if we choose that.

Reply:
> What `wl_surface` roles _should_ be supported? `xdg_toplevel`, and…? _Maybe_ layer shell surfaces, but they're already pretty special and I don't think any of this window management applies to them. `wl_shell` is dead dead dead and no client is going to use it in new code.

An alternative is to include the other archetypes Mir currently deduces from the `xdg-shell` role & attributes. And document how that happens. Then we need only talk about archetypes?

Layer-shell, is something orthogonal